### PR TITLE
EE-1079: enrich balance with merkle proofs, validate

### PR DESCRIPTION
**Pull request by henrytill**
Ref: https://casperlabs.atlassian.net/browse/EE-1079

Followed a similar approach to what we did with `query-state` here (using the purse lookup scheme as opposed to the query recursion), so nothing should be terribly surprising.

### CODE REVIEW

File: execution_engine/src/core/engine_state/balance.rs [hunk 1]
```diff
@@ -1,11 +1,43 @@
...
+    storage::trie::merkle_proof::TrieMerkleProof,
+};
 
 #[derive(Debug)]
 pub enum BalanceResult {
     RootNotFound,
-    Success(U512),
+    Success {
+        motes: U512,
+        main_purse_proof: Box<TrieMerkleProof<Key, StoredValue>>,
```
Comment (EdHastingsCasperLabs) [inline execution_engine/src/core/engine_state/balance.rs:13]: this should be `purse_proof` instead of `main_purse_proof`

**Review by EdHastingsCasperLabs — APPROVED** [review]
Other than requested name change, lgtm. Some nice adjacent improvements as well.

File: execution_engine/src/core/engine_state/balance.rs (hunk 1 above)
Comment (xcthulhu) [reply execution_engine/src/core/engine_state/balance.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/438/commits/49f9f4b5b9ec8ed59730b0ba65f4992b66597632

**Comment by henrytill** [comment]
one second...

**Comment by henrytill** [comment]
Too hasty again...

**Pull request merged**